gh-63416: Speed up assertEqual on long sequences

jdevries3133 opened the pull request:
This incorporates changes started by @eamanu:

- Fix [bpo-19217](https://bugs.python.org/issue19217)
- fix unnecessary space and solve @gpshead comments
- fix test
- [bpo-19217](https://bugs.python.org/issue19217): fix failing test in unittest's test suite
- [bpo-19217](https://bugs.python.org/issue19217): add blurb

Additionally, I added two commits to make this ready-to-merge

- revert changes to difflib.py
- add regression test

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- issue-number: [bpo-19217](https://bugs.python.org/issue19217) -->
https://bugs.python.org/issue19217
<!-- /issue-number -->


<!-- gh-issue-number: gh-63416 -->
* Issue: gh-63416
<!-- /gh-issue-number -->

eamanu committed:
Fix bpo-19217

eamanu committed:
fix unnecessary space and solve @gpshead comments

Commit:
fix test

jdevries3133 committed:
Merge remote-tracking branch 'temp/fix_bpo-19217' into bpo-19217-slow-assertEq

jdevries3133 committed:
bpo-19217: fix failing test in unittest's test suite

jdevries3133 committed:
bpo-19217: add blurb

jdevries3133 committed:
revert changes to difflib.py

jdevries3133 committed:
add regression test

ambv commented on Lib/unittest/test/test_case.py:
```diff
@@ -835,6 +839,36 @@ def testAssertSequenceEqualMaxDiff(self):
+        # test comparison of strings
+        arr1 = 'a' * 10000
+        arr2 = 'a' * 9999 + 'b'
+        start = time.time()
```
Always use `time.monotonic()` for timing comparisons.

ambv commented on Lib/unittest/test/test_case.py:
```diff
@@ -835,6 +839,36 @@ def testAssertSequenceEqualMaxDiff(self):
+            self.assertEqual(arr1, arr2)
+        except self.failureException:
+            pass
+        finish = time.time()
```
Ditto.

ambv commented on Lib/unittest/test/test_case.py:
```diff
@@ -835,6 +839,36 @@ def testAssertSequenceEqualMaxDiff(self):
+        except self.failureException:
+            pass
+        finish = time.time()
+        if finish - start > TIME_LIMIT:
```
I don't feel too good about this as it's hard to choose a good threshold for such checks.

In my particular case, at least on my laptop this passes no problem **without** any fix. And with almost 0.9s to spare! But then, if we tighten it too much, it will be failing on slower buildbots or when tests are run in parallel or under load.

My intuition is to drop the test altogether.

ambv commented on Lib/unittest/test/test_case.py:
```diff
@@ -835,6 +839,36 @@ def testAssertSequenceEqualMaxDiff(self):
+
+        # test comparison of strings
+        arr1 = 'a' * 10000
+        arr2 = 'a' * 9999 + 'b'
```
This is not a representative example. It results in exactly the same timings before and after the change **because it doesn't exercise the changed code** (which is in `assertSequenceEqual` and not in `assertMultiLineEqual`).

This makes me think we want to change `assertMultiLineEqual` as well since it's splitting the string into lines. This was mentioned by Trey Hunner on the issue.

ambv commented on Lib/unittest/test/test_case.py:
```diff
@@ -835,6 +839,36 @@ def testAssertSequenceEqualMaxDiff(self):
+        # test comparison of lists
+        arr1 =  [1] * 10000
+        arr2 = ([1] * 9999) + [2]
+        start = time.time()
```
This is again a non-representative example, taking the same time before and after the change. You need to generate a diff with more than 10 differences to see the difference in timing.

For example, if I use the following, there's a big difference:
```py3
        arr1 =  [1] * 1000
        a2 = ([1] * 99) + [2]
        arr2 = a2 * 10
```

New version takes 0.05s whereas the original takes a whopping 150s **and** requires raising the recursion limit for it to show the final diff 😂. And the increase is exponential, if you change the list size to 2,000 then the original takes a whopping 822s!

Sadly, [the old version looked quite good](https://gist.github.com/ambv/da471cb6e45beafa0068f4e5ceda00e8) while [the new one is pretty useless](https://gist.github.com/ambv/880e6cff441250d422e17f5fbf93c217). But that differs between cases so I guess it's better to get a bad diff than no diff in 15 minutes.

jdevries3133 replied:
I didn't think testing runtime was a good idea either, but I was hoping for an answer via Cunningham's Law...

Speaking of the recursion limit, maybe that's a way to get a better regression test? 

```python
height = len(inspect.stack(0))
original_limit = sys.getrecursionlimit()
sys.setrecursionlimit(height + <whatever we need>)
# do test
# restore recursion limit
```
What do you think?

jdevries3133 replied:
Nah, that idea is not good because as you said, the recursion only happens when the result is printed, not as it's being generated.

ambv replied:
Yeah, we don't need to touch the recursion limit just for this test. In fact, the passing version won't be exceeding the recursion limit anyway.

ambv reviewed (changes requested):
To sum up comments:
* Please add the new variant to `assertMultiLineEqual` as well;
* Please adapt the test cases to be pathological (you can use my suggestion) with the original code for both strings and lists;
* Please change the test to simply *execute* the diff, don't measure time at all. If it becomes too slow, we'll know just by the fact it became super slow.

jdevries3133 committed:
Merge branch 'main' of github.com:python/cpython into bpo-19217-slow-assertEq

jdevries3133 committed:
draft implementation of unittest.case._heuristic_diff

* Now, we switch between difflib.ndiff and difflib.unified_diff based on
  input size.
* Threshold is sort of arbitrary, but seems to be working in the limited
  test cases I've written
* The full test suite is passing, with only very minor tweaks to
  existing tests!

jdevries3133 committed:
fix: remove now-unused imports from test_case.py

ambv commented on Lib/unittest/case.py:
```diff
@@ -153,6 +154,36 @@ def _is_subtype(expected, basetype):
+    """
+
+    # @ambv: I just deduced this number from guess and check.... is there a
+    # better way?
```
I don't know a better way honestly. Test with as many diverse examples as you can would be my approach.

```suggestion
```

jdevries3133 replied:
Should it be exposed as a global constant, or kept inside the function?

ambv commented on Lib/unittest/test/test_case.py:
```diff
@@ -54,6 +54,126 @@ def tearDown(self):
+    def test_ndiff_to_unified_diff_breaking_point_varied_inputs(self):
+        """Specify how the heuristic behaves with varied inputs and edge cases."""
+        # @ambv: I just have this one scaffoleded out in comments. I'd like to
+        # know your thoughts in general before I keep going.
```
Looks good. We don't need to be exhaustive, we need to cover as many diverse examples as possible. Kind of go back to the origin of the term "edge case": look for cases when behavior is likely to change.

```suggestion
```

ambv commented on Lib/unittest/test/test_case.py:
```diff
@@ -54,6 +54,126 @@ def tearDown(self):
+
+    # @ambv: I feel that the tests between the lines here are redundant, given
+    # the other tests below. Removing them shaves ~1s of runtime off.
+    # what do you think?
```
1s doesn't matter.

```suggestion
```

ambv commented on Lib/unittest/test/test_case.py:
```diff
@@ -54,6 +54,126 @@ def tearDown(self):
             self.events.append('tearDown')
 
 
+class Test_HeuristicDiff(unittest.TestCase):
```
nit: I'd put this below the default test case.

ambv commented on Lib/unittest/test/test_case.py:
```diff
@@ -1,10 +1,10 @@
 import re
 import sys
 import logging
+from typing import Tuple
```
For Python 3.10+ we don't need this.

```suggestion
```

ambv commented on Lib/unittest/test/test_case.py:
```diff
@@ -54,6 +54,126 @@ def tearDown(self):
+
+    N = 50_000
+
+    def check(self, a, b, expect: Tuple[str, ...]):
```
```suggestion
    def check(self, a, b, expect: tuple[str, ...]):
```

ambv reviewed (commented):
I like where this is going. Looking forward to the next iteration.

jdevries3133 committed:
merge upstream python/cpython:main

jdevries3133 committed:
add variably scaled test cases, misc updates & revisions

jdevries3133 committed:
move Test_HeuristicDiff beneath main tests

ambv commented on Lib/unittest/case.py:
```diff
@@ -153,6 +154,37 @@ def _is_subtype(expected, basetype):
     return isinstance(expected, type) and issubclass(expected, basetype)
 
+
+def _heuristic_diff(a: list[str], b: list[str]) -> Iterable[str]:
```
You meant `Iterator`, not `Iterable`.

ambv commented on Lib/unittest/case.py:
```diff
@@ -153,6 +154,37 @@ def _is_subtype(expected, basetype):
+        cost = (number of differing lines
+                * total length of all differing lines)
+    """
+    # bpo-19217: speed up assertEqual on long sequences
```
```suggestion

        See bpo-19217 for context.
    """
```

ambv commented on Lib/unittest/case.py:
```diff
@@ -153,6 +154,37 @@ def _is_subtype(expected, basetype):
+    # of the differences, which is a proxy for the cost of difflib.ndiff
+    udiff = [l for l in difflib.unified_diff(a, b,
+                                             fromfile='expected',
+                                             tofile='got')]
```
```suggestion
    udiff = list(difflib.unified_diff(a, b, fromfile="expected", tofile="got"))
```

ambv commented on Lib/unittest/test/test_case.py:
```diff
@@ -1950,5 +1953,234 @@ def test2(self):
 
+class Test_HeuristicDiff(unittest.TestCase):
+
+    # this large contant coerces the use of `unified_diff` for several tests
```
```suggestion
    # this large constant coerces the use of `unified_diff` for several tests
```

jdevries3133 committed:
remove unnecessary list comprehension in Lib/unittest/case.py

Co-authored-by: Łukasz Langa <lukasz@langa.pl>

jdevries3133 committed:
spelling error in Lib/unittest/test/test_case.py

Co-authored-by: Łukasz Langa <lukasz@langa.pl>

jdevries3133 committed:
implement second review from @ambv

jdevries3133 committed:
merge changes from PR suggestions

JelleZijlstra commented on Lib/unittest/case.py:
```diff
@@ -153,6 +154,35 @@ def _is_subtype(expected, basetype):
+    # of the differences, which is a proxy for the cost of difflib.ndiff
+    udiff = list(difflib.unified_diff(a, b, fromfile="expected", tofile="got"))
+    udiff_differing_lines = [l for l in udiff
+                             if l.startswith('-') or l.startswith('+')]
```
```suggestion
                             if l.startswith(('-', '+'))]
```
And then the whole listcomp fits on one line.

JelleZijlstra commented on Lib/unittest/case.py:
```diff
@@ -5,6 +5,7 @@
 import difflib
 import pprint
 import re
+from typing import Iterator
```
```suggestion
from collections.abc import Iterator
```
typing.Iterator is deprecated

JelleZijlstra commented on Lib/unittest/case.py:
```diff
@@ -153,6 +154,35 @@ def _is_subtype(expected, basetype):
+    num_difflines = len(udiff_differing_lines)
+    total_diffline_length = sum(len(l) for l in udiff_differing_lines)
+
+    diff_cost = num_difflines * total_diffline_length
```
It might be worth commenting on why we're doing the cost calculation this way. On the face of it, it seems like we're basically including the line count twice. Is there something in the perf characteristics of the ndiff algorithm that makes this the right metric to use?

jdevries3133 replied:
I don't have a CS degree, and don't have a strong understanding of time complexity, which is why I'm not sure if I can explain why it works correctly, so I will try, but please check my work if you can :)

`unified_diff` calls `difflib.SequenceMatcher.get_opcodes`, and then transparently converts those opcodes into a diff. It looks to me like the time complexity is `O(a * b)`, given two strings, `a`, and `b` to compare. So, diffs are produced in linear time by comparing one line at a time.

`ndiff`, on the other hand, uses many of the higher-level features found in `difflib.Differ`. In particular, `difflib.Differ._fancy_replace` is probably where the time complexity balloons. It is documented as follows:

```
When replacing one block of lines with another, search the blocks
for *similar* lines; the best-matching pair (if any) is used as a
synch point, and intraline difference marking is done on the
similar pair. Lots of work, but often worth it.
```

Let's look at the time complexity of each frame independently, imagining that the functions it calls have zero-cost, so we can then sum up these complexities later.

| Func or Method | Time Complexity | Notes |
| -- | -- | -- |
| `difflib.ndiff` | `O(1)` | proxy to `difflib.Differ.compare` |
| `difflib.Differ.compare` | `O(a + b)` (ignoring calls inside the loop, which will be summed later) | `cruncher.get_opcodes` is O(a + b), and the size of the opcodes is linearly related to the extent of the diff, I believe |
| `difflib.Differ.dump` | `O(n)` | called for opcodes `delete`, `insert`, and `equal` |
| `difflib.Differ._fancy_replace` | `O(???)` –– a lot | I count at least 3 nested for loops in some edge cases. There is caching to take into account. I really don't have a clue – let's call it `O((a + b)^2)` on a good day. This is only called for `replace` opcodes, where it annotates the replacement for a better diff. Very pricy.

So, `ndiff` has an overall time complexity of:

```
O(a + b)    :: from calls to `get_opcodes`, and looping over those opcodes
* O(n)      :: this is the effect of calls to `Differ._dump` inside the loop in `Differ.compare`,
               which can effectively be ignored because it's not worse than `_fancy_replace`
* O((a + b)^2)    :: ...or more. This is the big nasty – calls to `_fancy_replace` inside the loop
               in `Differ.compare`
```

Therefore, the time complexity of `ndiff` is in the neighborhood of `O(n^2)`, I think? `unified_diff` seems closer to O(n).

I might be totally wrong, though... @JelleZijlstra what do you think?

jdevries3133 replied:
I did try to improve documentation of the source in b9f2f9db79d971a2022ad915d42198827aed7b18 based on my limited understanding.

jdevries3133 replied:
I did try to improve documentation of the source in b9f2f9db79d971a2022ad915d42198827aed7b18 based on my limited understanding.

JelleZijlstra replied:
Thanks, I think the text you added and the discussion here are very helpful! From a little IPython experimentation it does look like unified_diff is linear and ndiff is quadratic on a simple task (comparing a list of N "a" with N "b"):
```
In [19]: def nd(n):
    ...:     a = ["a"] * n
    ...:     b = ["b"] * n
    ...:     list(difflib.ndiff(a, b))
    ...: 
    ...: 

In [20]: def ud(n):
    ...:     a = ["a"] * n
    ...:     b = ["b"] * n
    ...:     list(difflib.unified_diff(a, b))
    ...: 
    ...: 

In [21]: %timeit ud(1000)
620 µs ± 12.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [24]: %timeit ud(10000)
6.96 ms ± 513 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [25]: %timeit ud(100000)
74.5 ms ± 9.38 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [26]: %timeit nd(100)
19.9 ms ± 638 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [27]: %timeit nd(1000)
1.89 s ± 135 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [28]: %timeit nd(10000)
```
For unified_diff, 10x as much data makes it ~10x slower, but for ndiff 10x the data makes it ~100x slower. I am not sure I am going to wait for that last one to finish.

jdevries3133 replied:
Ok cool, so it seems like what I added in b9f2f9d is accurate, then – I'm glad to hear it!

Do you think that b9f2f9d satisfies what you are looking for in terms of better documenting the heuristic?

JelleZijlstra replied:
Yes, this is good!

JelleZijlstra commented on Lib/unittest/test/test_case.py:
```diff
@@ -9,6 +10,7 @@
 import weakref
 import inspect
 import types
+from typing import Iterator
```
```suggestion
from collections.abc import Iterator
```

JelleZijlstra commented on Lib/unittest/test/test_case.py:
```diff
@@ -799,8 +801,9 @@ def testAssertSequenceEqualMaxDiff(self):
         seq2 = 'b' + 'x' * 80**2
-        diff = '\n'.join(difflib.ndiff(pprint.pformat(seq1).splitlines(),
-                                       pprint.pformat(seq2).splitlines()))
+        diff = unittest.case._heuristic_diff(pprint.pformat(seq1).splitlines(),  # type: ignore
```
```suggestion
        diff = unittest.case._heuristic_diff(pprint.pformat(seq1).splitlines(),
```
There's no type checker to ignore here. This reappears several times below.

JelleZijlstra commented on Lib/unittest/test/test_case.py:
```diff
@@ -1950,5 +1953,234 @@ def test2(self):
+                line_length_factor=1,
+                num_lines_factor=2,
+                extent_differing=1,
+                expect_unified_diff_at = 22,
```
```suggestion
                expect_unified_diff_at=22,
```
Consistent formatting (also reappears several times)

JelleZijlstra replied:
This one wasn't changed yet

JelleZijlstra replied:
@jdevries3133 does this give the right context? This conversation was hidden by default so maybe you missed it.

JelleZijlstra commented on Lib/unittest/test/test_case.py:
```diff
@@ -1950,5 +1953,234 @@ def test2(self):
+            given `N` constant value. Check if the test has passed as
+            specified."""
+
+            # we are working out way towards _heuristic_diff(foo, bar)
```
```suggestion
            # we are working our way towards _heuristic_diff(foo, bar)
```

JelleZijlstra commented on Lib/unittest/test/test_case.py:
```diff
@@ -1950,5 +1953,234 @@ def test2(self):
+                                 * int(N
+                                       * case.num_lines_factor
+                                       * (1 - case.extent_differing)))
+                bar = tuple((*bar_a_lines, *bar_b_lines, *bar_a_padding))
```
```suggestion
                bar = (*bar_a_lines, *bar_b_lines, *bar_a_padding)
```
It's already a tuple

JelleZijlstra commented on Lib/unittest/test/test_case.py:
```diff
@@ -1950,5 +1953,234 @@ def test2(self):
+                bar = tuple((*bar_a_lines, *bar_b_lines, *bar_a_padding))
+            else:
+                bar_b_lines = ((bar_b_line,) * int(N * case.num_lines_factor))
+                bar = tuple((*bar_a_lines, *bar_b_lines))
```
```suggestion
                bar = (*bar_a_lines, *bar_b_lines)
```

JelleZijlstra commented on Lib/unittest/test/test_case.py:
```diff
@@ -1950,5 +1953,234 @@ def test2(self):
+            # after all that, we have `foo` and `bar`; two string sequences
+            # with differences as specified by the Case parameters, scaled
+            # by a factor of `N`.
+            diff = unittest.case._heuristic_diff(foo, bar)  # type: ignore
```
```suggestion
            diff = unittest.case._heuristic_diff(foo, bar)
```

JelleZijlstra commented on Misc/NEWS.d/next/Library/2021-07-21-22-57-51.bpo-19217.vm-cr-.rst:
```diff
@@ -0,0 +1,2 @@
+Optimize unittest.TestCase.assertEqual method for long lists of varied
```
```suggestion
Optimize unittest.TestCase.assertEqual method for long sequences of varied
```

JelleZijlstra reviewed (changes requested):
Thanks, this is an exciting contribution! I have a bunch of small feedback. The bigger issue is that I think it's worth documenting in a bit more detail the background of the cost calculation and the perf characteristics of the two differs used.

arhadthedev commented on Lib/difflib.py:
```diff
@@ -1187,7 +1187,11 @@ def unified_diff(a, b, fromfile='', tofile='', fromfiledate='',
+            isinstance(b, list) or isinstance(b, tuple):
+            yield '{}'.format(lineterm)
+        else:
+            yield '@@ -{} +{} @@{}'.format(file1_range, file2_range, lineterm)
```
I propose to use interpolated strings to improve readability.

```suggestion
            yield f'{lineterm}'
        else:
            yield f'@@ -{file1_range} +{file2_range} @@{lineterm}'
```

jdevries3133 replied:
Fair enough, but that's outside the scope of this BPO, because I'm not making any changes in difflib, and I'm not otherwise making changes to this part of code.

jdevries3133 committed:
Merge branch 'main' of github.com:python/cpython into bpo-19217-slow-assertEq

jdevries3133 committed:
fix from @JelleZijlstra

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

jdevries3133 committed:
fix from @JelleZijlstra

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

jdevries3133 committed:
remove unnecessary type checker supression

jdevries3133 committed:
fix typo

jdevries3133 committed:
simplify tuple syntax

jdevries3133 committed:
fix news entry "~lists~ => sequenecs"

jdevries3133 committed:
better document the reasoning behind the heuristic

JelleZijlstra commented on Lib/unittest/case.py:
```diff
@@ -170,6 +171,35 @@ def _is_subtype(expected, basetype):
+def _heuristic_diff(a: list[str], b: list[str]) -> Iterator[str]:
+    """After testing the magnitude of the inputs, preferably return the output
+    of difflib.ndiff, but fallback to difflib.unified_diff for prohibitively
+    expensive inputs. How cost is calclated:
```
```suggestion
    expensive inputs. How cost is calculated:
```

JelleZijlstra commented on Misc/NEWS.d/next/Library/2021-07-21-22-57-51.bpo-19217.vm-cr-.rst:
```diff
@@ -0,0 +1,2 @@
+Optimize unittest.TestCase.assertEqual method for long sequences of varied
```
```suggestion
Optimize :meth:`~unittest.TestCase.assertEqual` method for long sequences of varied
```
Link to the unittest docs

JelleZijlstra reviewed (changes requested):
Thanks! I have a couple of tiny comments

jdevries3133 committed:
thanks @JelleZijlstra

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

jdevries3133 committed:
thanks @JelleZijlstra

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

jdevries3133 committed:
fix whitespace around keyword argument

jdevries3133 committed:
Merge branch 'bpo-19217-slow-assertEq' of github.com:jdevries3133/cpython into bpo-19217-slow-assertEq

JelleZijlstra reviewed (approved):
Thanks, this is great! Once main is no longer blocked I can help find someone who can merge this PR :)

One more thing to consider: Should this be backported to the 3.9 and 3.10 branches? I think we generally don't backport performance optimizations, so I'm leaning towards no.

gpshead commented on Lib/unittest/test/test_case.py:
```diff
@@ -1973,5 +1976,234 @@ def test2(self):
+                          'switch at {case.expect_unified_diff_at}, but no '
+                          f'switch occured when N == {N}')
+
+        for case in cases:
```
You're basically implementing test case parameterization here within one test. I woefully miss things like https://github.com/abseil/abseil-py/blob/main/absl/testing/parameterized.py or pytest's parameterized in the stdlib...

anyways, not a big deal.  this will work.  but a future refactoring when we support some form of easy parameterization in the stdlib would generate ~100 different test methods out of this. :)

jdevries3133 replied:
Yes, I see patterns like this all over the cpython regression test suite – not sure if there's a better way with the tools currently available. I just tried to make it as clear as possible with the dataclass and everything!

gpshead commented on Misc/NEWS.d/next/Library/2021-07-21-22-57-51.bpo-19217.vm-cr-.rst:
```diff
@@ -0,0 +1,2 @@
+Optimize :meth:`~unittest.TestCase.assertEqual` method for long sequences of varied
```
Mention what observable change users will notice.  Faster assertion reporting, but also presumably differently formatted diffs in formerly slow cases?

jdevries3133 replied:
Ok, I updated it in fc752cf40b752a28781e67b457de0d31f54dee14. Let me know if you have any thoughts.

jdevries3133 committed:
update blurb to describe observable changes

jdevries3133 committed:
Merge branch 'main' of github.com:python/cpython into bpo-19217-slow-assertEq

tim-one reviewed (commented):
I haven't changed my mind since the last time I commented on one of the related issue reports (which was some years ago): if unittest is determined to compare strings, regardless of the types of the arguments passed to `assertEqual()`, then it should define the way it constructs such strings, and use its own dirt-simple, worst-case linear-time string diff algorithm. Just, e.g., look for the first mismatching character.

You can try to out-think difflib, but at best you'll make "the problem" less likely. If that's good enough for you, fine by me :wink:. The approach here, to my eyes, should work pretty well for most cases.

Note that people here are generally comparing the very best case for unified diff to the very worst case for ndiff: no strings in common. Unified diff does only one layer of differencing, and if no elements are in common it has next to nothing to do. But ndiff does two layers: if there are no strings in common, it goes on to compare every string in one argument to every string in the other, to find a pair "closest to" matching.

But unified diff _can_ be quadratic time too, when there are lots of matching subsequences.

It's convoluted to construct such cases, because the `autojunk=True` default works against it (but also damages the quality of the diffs). Here's a simple demonstration by forcing `autojunk=False`, using so-called "Fibonacci strings", which are a worst case for many string-matching algorithms:

```py
from difflib import SequenceMatcher
fs = ['a', 'b']
from time import perf_counter as now
while True:
    a = fs[-2]
    b = fs[-1]
    start = now()
    r = SequenceMatcher(None, a, b, autojunk=False).ratio()
    finish = now()
    print(len(b), finish - start)
    fs.append(a + b)
```

The much-worse-than-linear behavior then quickly becomes obvious at non-trivial, but still smallish, string lengths:

```
...
144 0.0011279999998805579
233 0.0033622999999352032
377 0.008270400000583322
610 0.027310899999974936
987 0.05379209999955492
1597 0.1655982000002041
2584 0.4110897999998997
4181 1.1430293000003076
6765 2.670060499999636
10946 7.549530700000105
17711 19.679138700000294
28657 51.00185919999967
46368 137.25895500000024
75025 383.6125170999994
...
```

That was done on a far-from-quiet machine, but the point should be obvious enough.

gpshead committed:
reword to make less promises

unified diff is not "fast" just "less likely slow" or "slow in different circumstances".  so this doesn't ensure the problem never happens, it just reduces its chance.

gpshead reviewed (dismissed):
I tweaked the wording in the NEWS entry a little.

tim-one commented on Lib/unittest/case.py:
```diff
@@ -170,6 +171,46 @@ def _is_subtype(expected, basetype):
+    approximately O((diff)^2), where `diff` is the product of the number of
+    differing lines, and the total length of differing lines. On the other
+    hand, unified_diff's cost is the same as the cost of producing `diff`
+    by itself: O(a + b).
```
Note that this isn't true. Both have worst-case quadratic time, but on different kinds of input, and it's less _likely_ to get provoked in the simpler kind of differencing `unified_diff` tries.

Well, actually actually :wink:, the possible worst case of `ndiff` is worse than that, at least cubic time.

gpshead reviewed (commented):
It sounds like we're basically trading one regressive performance situation for another. 

Someone can always wind up unhappy. Without the ability to bail out if computation is taking too long, we should take Tim's suggestion of going to a linear "first differing element" highlight approach when inputs are "too large", followed by our generic a != b printing code.

I blindly suggest: `max(len(a), len(b)) > 500` prevent calling unified_diff and `max(len(a), len(b)) > 200` prevents calling ndiff. Just based on the worst case timings.

those default values used to tune that could be TestCase attributes that people could override themselves if they want (assumed to be rare), much like maxDiff is today.

jdevries3133 committed:
Merge branch 'main' of github.com:python/cpython into bpo-19217-slow-assertEq

jdevries3133 committed:
Merge branch 'bpo-19217-slow-assertEq' of github.com:jdevries3133/cpython into bpo-19217-slow-assertEq

iritkatriel reviewed (changes requested):
This has merge conflicts now.

AA-Turner committed:
Merge remote-tracking branch 'upstream/main' into bpo-19217-slow-assertEq

AA-Turner committed:
Remove annotations

AA-Turner commented on Lib/unittest/case.py:
```diff
@@ -190,6 +190,46 @@ def _is_subtype(expected, basetype):
+
+    # call unified_diff
+    udiff = list(difflib.unified_diff(a, b, fromfile="expected", tofile="got"))
+    udiff_differing_lines = [l for l in udiff
+                             if l.startswith(('-', '+'))]
```
```suggestion
    udiff_differing_lines = [l for l in udiff if l.startswith(('-', '+'))]
```

AA-Turner commented on Lib/unittest/case.py:
```diff
@@ -190,6 +190,46 @@ def _is_subtype(expected, basetype):
+
+    # inspect unified_diff output
+    num_difflines = len(udiff_differing_lines)
+    total_diffline_length = sum(len(l) for l in udiff_differing_lines)
```
```suggestion
    total_diffline_length = sum(map(len, udiff_differing_lines))
```

AA-Turner commented on Lib/unittest/case.py:
```diff
@@ -190,6 +190,46 @@ def _is_subtype(expected, basetype):
+    hand, unified_diff's cost is the same as the cost of producing `diff`
+    by itself: O(a + b).
+
+    See bpo-19217 for additional context.
```
```suggestion
    See gh-63416 for additional context.
```
